### PR TITLE
Pr/2390

### PR DESCRIPTION
**Pull request by qwwdfsad**

### CODE REVIEW

File: kotlinx-coroutines-core/jvm/test/sync/SemaphoreStressTest.kt [hunk 1]
```diff
@@ -90,4 +90,22 @@ class SemaphoreStressTest : TestBase() {
...
+        val semaphore = Semaphore(1)
+        val n = 1000 * stressTestMultiplier
+        repeat(n) {
+            val job = launch(Dispatchers.Default) {
+                semaphore.acquire()
+                semaphore.release()
+            }
+            semaphore.withPermit {
+                job.cancel()
+
```
Comment (elizarov) [inline kotlinx-coroutines-core/jvm/test/sync/SemaphoreStressTest.kt:105]: Nit: extra empty line

File: kotlinx-coroutines-core/common/src/sync/Mutex.kt [hunk 2]
```diff
@@ -201,7 +201,8 @@ internal class MutexImpl(locked: Boolean) : Mutex, SelectClause2<Any?, Mutex> {
                         // try lock
                         val update = if (owner == null) EMPTY_LOCKED else Empty(owner)
                         if (_state.compareAndSet(state, update)) { // locked
-                            cont.resume(Unit)
+                            // TODO implement functional type in LockCont as soon as we get rid of legacy JS
```
Comment (elizarov) [inline kotlinx-coroutines-core/common/src/sync/Mutex.kt:204]: Btw, if that helps, we can optimize it for `onwer==null` case by caching a reference to a preallocated `onCancellationUnlock` reference similar to the way it is done in Semaphore.

**Review by elizarov — APPROVED** [review]
💯 Awesome. That's the right way to fix it. Tests are also great. See some comments.

File: kotlinx-coroutines-core/common/src/sync/Mutex.kt [hunk 3]
```diff
@@ -201,7 +201,8 @@ internal class MutexImpl(locked: Boolean) : Mutex, SelectClause2<Any?, Mutex> {
...
+                            cont.resume(Unit) { unlock(owner) }
```
Comment (yschimke) [inline kotlinx-coroutines-core/common/src/sync/Mutex.kt:205]: I love Kotlin's last param is lambda logic generally. I think in this API, this callsite is an example of where it hides the intent.

```
cont.resume(Unit, onCancellation = { unlock(owner) })
```

**Pull request merged**